app/engine/penguin.py

```diff
@@ -140,38 +139,65 @@ def initialize_power_cards(self, session=None) -> None:
             'fire': 'f'
         }[self.element]
 
-        power_cards = cards.fetch_power_cards_by_penguin_id(
-            self.pid,
-            element_name,
-            session=session
-        )
+        power_cards = None
+
+        if self.is_bot:
+            power_cards = cards.fetch_by_element(
+                element_name, 
+                session=session
+            )
+
+        else:
+            power_cards = cards.fetch_power_cards_by_penguin_id(
+                self.pid,
+                element_name,
+                session=session
+            )
 
         for card in power_cards:
             card_object = CardObject(card, self)
             card_object.color = card_color
-            self.power_cards.append(card_object)
+            self.owned_cards.append(card_object)
+
+        self.logger.info(f"Initializing {len(self.owned_cards)} cards for {self.element}")
 
     def next_power_card(self) -> CardObject | None:
-        if not self.power_cards:
+        if not self.has_power_cards:
             # Client has no more power cards
+            self.logger.info(f'{self.name} has no more power cards')
             return
 
-        if len(self.power_card_slots) >= 4:
-            # Client cannot hold more than 4 power cards
-            return
+        next_card = random.choice(self.owned_cards)
+        self.owned_cards.append(next_card)
+        self.owned_cards.remove(next_card)
 
-        next_card = random.choice(self.power_cards)
-        self.power_card_slots.append(next_card)
-        self.power_cards.remove(next_card)
         return next_card
 
     def power_card_by_id(self, card_id: int) -> Card | None:
-        return next((c for c in self.power_card_slots if c.id == card_id), None)
+        return next((c for c in self.owned_cards if c.id == card_id), None)
 
     def update_cards(self) -> None:
         if self.disconnected:
             return
 
+        if self.is_bot:
+            return
+
+    def update_cards(self) -> None:
+        if self.disconnected:
+            return
+
+        if self.is_bot:
+            return
+
+        if not self.has_power_cards:
+            self.logger.info(f'{self.name} needs to purchase more cards from the catalogue!')
+            return
+
+        if self.cards_ready >= 4:
+            self.logger.info(f'{self.name} needs to place a card before queuing a new one!')
+            return
```

Comment:
Why is this there? This should get handled below, with the `cycle` update attribute:
```python
if len(self.owned_cards) > 3:
    update['cycle'] = True
```

Reply:
Could this check be added to the `next_power_card` function, like this?
```python
def next_power_card(self) -> CardObject | None:
    if not self.has_power_cards:
        # Client has no more power cards
        self.logger.info(f'{self.name} has no more power cards')
        return

  if self.cards_ready >= 4:
      self.logger.info(f'{self.name} needs to place a card before queuing a new one!')
      return

    next_card = random.choice(self.owned_cards)
    self.owned_cards.append(next_card)
    self.owned_cards.remove(next_card)

    return next_card
```

because this would mean that the stamina bar still goes up, like it used to.

Reply:
I thought update['cycle'] = True determines visible cycling of the card queue shown in the UI when sent as a payload to the client?

Also, I don't see why we are appending to the end of self.owned_cards and then removing the same entry?
Wouldn't this leave the list as it was?

Reply:
The `.append` & `.remove` was there in [your original code](https://github.com/Walinski/snowflake2/blob/67c64d4b407d3753d0dd630ccf2e45a584d4a62e/app/engine/penguin.py#L165), so I am not sure if this is intended.

The stamina bar is supposed to go up, even if you have all slots filled up already. But don't worry, this doesn't mean they "waste" anything. This is where the `cycle` attribute comes in. This one just resets the stamina back to 0, without giving you any card, so that it can go back up again. If you want I could try to find a video for this, but its very rare to see someone having 4 powercards at once.

![image](https://github.com/user-attachments/assets/143ea438-b350-4045-8c9a-5e4318f2df07)